rpcserver: Create Chain and UtxoEntry interfaces.

rstaudt2 opened the pull request:
This creates rpcserver.Chain and rpcserver.UtxoEntry interfaces and
respective rpcChain and rpcUtxoEntry adaptors that implement those
interfaces.

This starts to decouple rpcserver from the concrete BlockChain
implementation and allows for mocking those interfaces when testing
rpcserver handlers.

This addresses https://github.com/decred/dcrd/issues/2069 and adds test coverage for the following rpcserver handlers:
- handleCreateRawSSRtx
- handleCreateRawSStx
- handleDebugLevel
- handleDecodeRawTransaction
- handleExistsExpiredTickets
- handleExistsLiveTicket
- handleExistsLiveTickets
- handleExistsMissedTickets
- handleGetBestBlock
- handleGetBestBlockHash
- handleGetBlockCount
- handleGetChainTips
- handleGetCoinSupply
- handleGetTxOutSetInfo

rstaudt2 commented on internal/rpcserver/interface.go:
```diff
@@ -134,3 +138,300 @@ type SyncManager interface {
+	//
+	// The function is safe for concurrent access.
+	TipGeneration() ([]chainhash.Hash, error)
+}
```
This seems pretty verbose since rpcserver relies on quite a few blockchain methods, but also seems like the clearest (and more explicit) way to outline what rpcserver currently uses while also allowing for mocking - open to suggestions if anyone has a better idea

rstaudt2 replied:
I meant more just defining that many methods in the interface that align one-to-one with the blockchain methods - vs potentially breaking this down into smaller interfaces and/or consolidating some of the interfaces to be more geared for what rpcserver needs.

One issue with not commenting the interface would be that most editors have plugins that allow for peeking/viewing method comments on hover, so you would lose that ability

JoeGruffins replied:
I gotcha. imo this is about testing the handlers, and it seems you found a bug, so goal achieved. Optimization can come later, if needed.

Good point about the editors. That's something I hadn't thought of. Let's keep the comments then.

davecgh commented:
Thanks for the PR.  It's pretty large, so I won't be able to review it right away, but it's not being ignored.  Hopefully some others can help review it in the mean time as well.

JoeGruffins commented on rpcadaptors.go:
```diff
@@ -324,3 +328,405 @@ func (b *rpcSyncMgr) ProcessTransaction(tx *dcrutil.Tx, allowOrphans bool,
+}
+
+// Ensure rpcUtxoEntry implements the rpcserver.UtxoEntry interface.
+var _ rpcserver.UtxoEntry = &rpcUtxoEntry{}
```
I've never seen it done this way. I believe the idiomatic way to check that something satisfies an interface is so: `var _  rpcserver.UtxoEntry= (*rpcUtxoEntry)(nil)`

rstaudt2 replied:
Ya, good point - I followed what was already being done in that file for the other adaptors, which works, but I believe the way its currently being done will allocate memory for an empty struct whereas the way you outlined would not. I'll update those across this file

JoeGruffins commented on rpcadaptors.go:
```diff
@@ -324,3 +328,405 @@ func (b *rpcSyncMgr) ProcessTransaction(tx *dcrutil.Tx, allowOrphans bool,
+}
+
+// Ensure rpcChain implements the rpcserver.Chain interface.
+var _ rpcserver.Chain = &rpcChain{}
```
Here too.

JoeGruffins commented on rpcserver.go:
```diff
@@ -731,7 +731,7 @@ func handleCreateRawSStx(_ context.Context, s *rpcServer, cmd interface{}) (inte
 		if !(input.Tree == wire.TxTreeRegular ||
 			input.Tree == wire.TxTreeStake) {
-			rpcInvalidError("Tx tree must be regular or stake")
+			return nil, rpcInvalidError("Tx tree must be regular or stake")
```
Nice.

JoeGruffins commented on rpcserverhandlers_test.go:
```diff
@@ -6,28 +6,821 @@ package main
+	switch c.testTxHash.String() {
+	case "3189cbe656c2ef1e0fcb91f107624d9aa8f0db7b28e6a86f694a4cf49abc5e39":
+		return []*stake.MinimalOutput{{
+			PkScript: []byte{0xBA, 0xA9, 0x14, 0x78, 0x02, 0x39, 0xEA, 0x12, 0x31, 0xBA, 0x67, 0xB0, 0xC5, 0xB8, 0x2E, 0x78, 0x6B, 0x51, 0xE2, 0x10, 0x72, 0x52, 0x21, 0x87},
```
Could you break up these long lines? Here and in other places throughout the tests.

rstaudt2 replied:
Sure, I will adjust that. Thanks for your comments!

rstaudt2 commented:
@JoeGruffins Made adjustments per your comments so far in [this commit](https://github.com/decred/dcrd/pull/2211/commits/f312df7a1dc8325e0e9ea898007353e00fb83809)

JoeGruffins commented on rpcserverhandlers_test.go:
```diff
@@ -146,22 +979,472 @@ func TestHandleCreateRawTransaction(t *testing.T) {
+			},
+		},
+		wantErr: true,
+		errCode: dcrjson.ErrRPCInvalidParameter,
```
Same as `handleExistsExpiredTickets` and `handleExistsLiveTickets`

rstaudt2 commented on rpcserverhandlers_test.go:
```diff
@@ -6,28 +6,854 @@ package main
 // testServer is a rpcServer stub used for testing handlers.
 var testServer = &rpcServer{cfg: testCfg}
 
+type rpcTest struct {
+	name    string
+	handler commandHandler
+	cmd     interface{}
+	result  interface{}
+	wantErr bool
+	errCode dcrjson.RPCErrorCode
+}
```
Yep, I noticed that as well - but in this case, the struct type is used across all tests, so I thought it made more sense to define the type once rather than defining the type inline in every test.

davecgh replied:
I do very much prefer inline test structs for consistency, but there are a couple of other instances where it isn't true due to the same reason as here.  Given every one of these tests are identical, it isn't unreasonable to have the struct defined separately in this case.

rstaudt2 commented on rpcserverhandlers_test.go:
```diff
@@ -6,28 +6,854 @@ package main
+		},
+		wantErr: true,
+		errCode: dcrjson.ErrRPCInvalidParameter,
+	}, {
```
Regarding this comment and a couple of others below - yes, there are a few error scenarios that are unreachable, primarily due to the given rpcserver handler already checking for the same error scenario(s) that are handled in the method that it is calling. If we still want to test it, we could stub the method being called to return an error in these cases (since it can't be hit with just bad input parameters).

Since these are unreachable, we could ignore errors returned by the method being called in these cases, but that would make this code more coupled to the internals of that method (i.e., if a new error case was added, then the rpcserver code would need to be updated).

rstaudt2 commented:
@JoeGruffins Thanks for the feedback and taking the time to go through this PR. 

For this comment 
> I would argue that the `testRPCChain` struct should be more like the `testRPCUtxoEntry` and return values stored in the struct

I see your point - I originally went this route so that the mocked chain data wouldn't have to be defined in each test case, but looking back at it now I think I could add handling so that you could mock just the applicable methods inline in the test definition (and probably still have default values defined for the chain data so that the base case doesn't need to be copied across lots of tests).

rstaudt2 commented:
@JoeGruffins In the latest [commit](https://github.com/decred/dcrd/pull/2211/commits/77a1d03774c4b4ca65cfe782b6546f7ea9203549):

- Moved the mocking of the Chain methods to be inline within the tests - I think this does make it quite a bit easier to follow
- Made the corrections you pointed out in a few of the method comments
- Broke up the long hex strings

For your comment 
> Throughout the tests, comments are missing a full stop.

Do you mean that you think additional comments are needed within each test case? I thought given the `name`, parameters, and now mocked chain method details that it is pretty clear - but I can add something more verbose if needed.

JoeGruffins commented:
> Do you mean that you think additional comments are needed within each test case? I thought given the name, parameters, and now mocked chain method details that it is pretty clear - but I can add something more verbose if needed.

No. The comments are missing punctuation. No biggie tho.
For example, line 24:
`// testRPCUtxoEntry provides a mock utxo entry by implementing the UtxoEntry interface`
is a sentence so should have a period.

davecgh commented on internal/rpcserver/interface.go:
```diff
@@ -134,3 +138,300 @@ type SyncManager interface {
+	// TicketPoolValue returns the current value of all the locked funds in the
+	// ticket pool.
+	//
+	// This function is safe for concurrent access.  All live tickets are at least
+	// 256 blocks deep on mainnet, so the UTXO set should generally always have
+	// the asked for transactions.
```
That entire blurb should just be removed from here.  It isn't providing any relevant information to the caller and there is nothing they can due to influence it anyway.

davecgh commented on rpcadaptors.go:
```diff
@@ -324,3 +328,405 @@ func (b *rpcSyncMgr) ProcessTransaction(tx *dcrutil.Tx, allowOrphans bool,
+// rpcUtxoEntry represents a utxo entry for use with the RPC server and
+// implements the rpcserver.UtxoEntry interface.
+type rpcUtxoEntry struct {
+	utxoEntry *blockchain.UtxoEntry
```
Since the interface essentially is already implemented by `blockchain.UtxoEntry`, you can avoid almost all of this code in here by simply embedding the type directly.  Then you'd only need to define the `ToUtxoEntry()` case since all the others would automatically be inherited.

e.g. The following will implement it properly:

```Go
// rpcUtxoEntry represents a utxo entry for use with the RPC server and
// implements the rpcserver.UtxoEntry interface.
type rpcUtxoEntry struct {
	*blockchain.UtxoEntry
}

// Ensure rpcUtxoEntry implements the rpcserver.UtxoEntry interface.
var _ rpcserver.UtxoEntry = (*rpcUtxoEntry)(nil)

// ToUtxoEntry returns the underlying UtxoEntry instance.
func (u *rpcUtxoEntry) ToUtxoEntry() *blockchain.UtxoEntry {
	return u.UtxoEntry
}
```

Due to this, I'm skipping reviewing all of the other unnecessary funcs in this file since they'll be removed.

davecgh commented on rpcadaptors.go:
```diff
@@ -324,3 +328,405 @@ func (b *rpcSyncMgr) ProcessTransaction(tx *dcrutil.Tx, allowOrphans bool,
+// rpcChain provides a chain for use with the RPC server and
+// implements the rpcserver.Chain interface.
+type rpcChain struct {
+	chain *blockchain.BlockChain
```
The same thing is true here.  You can avoid hundreds of lines of code here by just embedding the type directly and defining only the cases that either don't exist or don't have matching signatures.  Namely, `ConvertUtxosToMinimalOutputs ` and `FetchUtxoEntry`.

e.g. The following will implement it properly:

```Go
// rpcChain provides a chain for use with the RPC server and
// implements the rpcserver.Chain interface.
type rpcChain struct {
	*blockchain.BlockChain
}

// Ensure rpcChain implements the rpcserver.Chain interface.
var _ rpcserver.Chain = (*rpcChain)(nil)

// ConvertUtxosToMinimalOutputs converts the contents of a UTX to a series of
// minimal outputs. It does this so that these can be passed to stake subpackage
// functions, where they will be evaluated for correctness.
func (c *rpcChain) ConvertUtxosToMinimalOutputs(entry rpcserver.UtxoEntry) []*stake.MinimalOutput {
	return blockchain.ConvertUtxosToMinimalOutputs(entry.ToUtxoEntry())
}

// FetchUtxoEntry loads and returns the unspent transaction output entry for the
// passed hash from the point of view of the end of the main chain.
//
// NOTE: Requesting a hash for which there is no data will NOT return an error.
// Instead both the entry and the error will be nil.  This is done to allow
// pruning of fully spent transactions.  In practice this means the caller must
// check if the returned entry is nil before invoking methods on it.
//
// This function is safe for concurrent access however the returned entry (if
// any) is NOT.
func (c *rpcChain) FetchUtxoEntry(txHash *chainhash.Hash) (rpcserver.UtxoEntry, error) {
	utxo, err := c.BlockChain.FetchUtxoEntry(txHash)
	return &rpcUtxoEntry{utxo}, err
}
```

rstaudt2 replied:
Makes sense, updated - I started this way to avoid exposing the underlying `blockchain.BlockChain` and its methods that are unrelated to what `rpcserver` requires. But since `rpcserver` is just using the `rpcserver.Chain` interface, and not the `rpcChain` adaptor directly, it doesn't matter.

rstaudt2 commented on internal/rpcserver/interface.go:
```diff
@@ -134,3 +138,300 @@ type SyncManager interface {
+	// Returns true if the output index references an output that does not exist
+	// either due to it being invalid or because the output is not part of the view
+	// due to previously being spent/pruned.
+	IsOutputSpent(outputIndex uint32) bool
```
Makes sense - thanks for the heads up. I'll think through that a bit more and see if it makes sense to bring in any changes to this PR or handle separately

rstaudt2 commented on internal/rpcserver/interface.go:
```diff
@@ -134,3 +138,300 @@ type SyncManager interface {
 		allowHighFees bool, tag mempool.Tag) ([]*dcrutil.Tx, error)
 }
+
+// UtxoEntry represents a utxo entry for use with the RPC server.
```
👍 I originally left it out since it is NOT required to be safe for concurrent access (as the underlying implementation of `blockchain.UtxoEntry` is not), but I updated this to state that explicitly now.

rstaudt2 commented:
@davecgh Thanks for reviewing. I'll look more into your comment regarding `UtxoEntry` being changed to an outpoint in the future and will also take another pass through the interface method comments. The rest of your feedback has been addressed in the latest commit.

davecgh commented on rpcserverhandlers_test.go:
```diff
@@ -6,28 +6,829 @@ package main
+	modified   bool
+}
+
+// TransactionType returns the txType of the testRPCUtxoEntry
```
I see this was fixed in the latest push while I was doing the review.  Resolving.

davecgh commented on rpcserverhandlers_test.go:
```diff
@@ -6,28 +6,829 @@ package main
-var testCfg = rpcserverConfig{ChainParams: chaincfg.MainNetParams()}
+// testRPCUtxoEntry provides a mock utxo entry by implementing the UtxoEntry interface
+type testRPCUtxoEntry struct {
+	rpcUtxoEntry
```
Right, that is what I figured was the case.  I'm not overly concerned with it the way it is if you prefer to leave it, since this is test code, but I did want to point out that it is a bit loose in terms of correctness and will result in panics if new handlers are added without overriding the relevant methods.

That said, a panic versus an explicit "not implement yet" error in the tests really doesn't make much difference when it comes to test failure since it's a failure either way.  It might be a little easier on someone less familiar with the code trying to implement tests for a new handler to see a "Foo is not implemented yet" test failure as opposed to a panic, but it should be pretty easy to figure out either way.

rstaudt2 commented:
@davecgh Resolved the remaining open comments you had:
- Adjusted `testRPCUtxoEntry` and `testRPCChain` to remove the embedding and fully implement the interfaces instead
- Fixed the duplicate test name `invalid amount < 0`
- Also updated a few of the longer test functions (such as `TestHandleCreateRawSStx` [here](https://github.com/decred/dcrd/pull/2211/files#diff-396c6be7acf6983648f03925639c97deR350-R362)) to remove some of the repetitive/verbose test details

davecgh commented:
@JoeGruffins This looks good to go from my perspective.  Is there anything outstanding you want addressed?
